### PR TITLE
Minimal working Savestates

### PR DESCRIPTION
**Pull request by m4xw**
TODO:
- Check audio (partly unserialized)
- Check cart interaction after state load (missing serialization of pos etc)
- Use fixed size Integer types for safe cross platform use
- Maybe serialize the frame contents & scanline states
- Cleanup erroneous replaces of PC/SP and others in comments

WIP PR to discuss, not ready for a merge yet.

### CODE REVIEW

**Comment by m4xw** [comment]
If theres no objections, then I will move on with cleanup and merge

**Comment by BlazeEntertainment** [comment]
Green light here Max. Please merge

**Review by Jamiras — COMMENTED** [review]
Tested briefly with Diner!

I created two save states at the title screen and two save states while the instructions were on-screen.
* Switching between the two title screen states worked fine.
* Switching between the two instruction states worked fine.
* Switching from a title screen state to an instruction state had garbled graphics.
* Switching from an instruction state to a title screen state had the wrong colors.

I'm guessing some of the `stic` state needs to be captured.

When using states within the game, I didn't notice any issues with sounds, so whatever "partially unserialized" data is there appears to be sufficient.

File: src/cp1610.h [hunk 1]
```diff
@@ -17,10 +17,36 @@
 	along with FreeIntv.  If not, see http://www.gnu.org/licenses/
 */
 
+#include <stdint.h>
+typedef struct CP1610_Context_t
+{
+	uint32_t Version;
+	int (*OpCodes[0x400])(int);
+	int Interuptable[0x400];
+	const char *Nmemonic[0x400];
```
[the remark this answers is not in the document]
Comment (m4xw) [reply src/cp1610.h:26]: Hey, thanks for review. Good call on that!
Consider them a oversight, at least for nmemonic i missed the "*", I will double check whats appropriate

```
Switching from a title screen state to an instruction state had garbled graphics.
Switching from an instruction state to a title screen state had the wrong colors.
```

Did that fix itself after another transition? I have a idea where thats coming from at least..

[the remark this answers is not in the document]
Comment (Jamiras) [reply src/cp1610.h:26]: > Did that fix itself after another transition? I have a idea where thats coming from at least..

Transitioning to another of the same type shows the same problem. Transitioning back to the original type corrects the issue.

* State1, State2 = title screen
* State3, State4 = instructions

Title screen -> State1 (ok) -> State2 (ok) -> State1 (ok)
Title screen -> State1 (ok) -> State2 (ok) -> State3 (garbled) -> State4(garbled) -> State1 (ok)
Title screen -> State3 (garbled) -> State1 (ok)
Instructions -> State1 (wrong colors)
Instructions -> State3 (ok) -> State4(ok) -> State3 (ok)
Instructions -> State3 (ok) -> State4(ok) -> State1 (wrong colors) -> State2 (wrong colors) -> State3 (ok)
Title screen -> State1 (ok) -> Wait for instructions -> State3 (ok) -> State1 (wrong colors)

Examples of correct (left) and incorrect (right) after loading states 1 and 3:
image

[the remark this answers is not in the document]
Comment (m4xw) [reply src/cp1610.h]: Can you check my latest push?

**Comment by BlazeEntertainment** [comment]
Checked with titles:

- Buzz Bombers 
- Word Rockets

Commits fixed the serialization issue and from what I can tell transitions seem OK but didn't tested DINER!

File: src/cp1610.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jamiras) [reply src/cp1610.h]: Looks good with the Diner! test cases that I was using.

[the remark this answers is not in the document]
Comment (m4xw) [reply src/cp1610.h]: Cool, i will merge it tomorrow then.

**Comment by m4xw** [comment]
I will consider this superseeded.

**Pull request closed without merging**